### PR TITLE
Complete prior Earley predictions

### LINKED ISSUE #117: Nullable productions (Still, Still) Fail to Match
Related to #108 and #112 – failures reproduce on `0.4.0 ≤ bnf ≤ 0.4.3` so _not_ a regression from #113 .

Hit more cases where nullable productions aren't handled properly.

This example looks contrived but was adapted from a larger grammar I'm writing.

As usual, thanks for all your work @CrockAgile !

Sadly, this correctness issue continues to be a total blocker preventing me from adopting `bnf`

Hoping my battle-testing hasn't been in vain and there's an easy fix around the corner 🙏🏻

```rs
use bnf::Grammar;

#[test]
fn test() {
    let fails: &str = "
    <disjunction> ::= <predicate> | <disjunction> <or> <predicate>
    <predicate> ::= <char_null_one> | <special-string> '.'

    <char_null_one> ::= <char_null_two>
    <char_null_two> ::= <char_null_three>
    <char_null_three> ::= <char>

    <or> ::= <ws> 'or' <ws>
    <ws> ::= <whitespace> | ' ' <ws>
    <whitespace> ::= ' '

    <special-string> ::= <special-char> | <special-char> <special-string>
    <special-char> ::= <char> | <whitespace>
    <char> ::= 'a'
    ";

    let input = "a or a";

    let passes_1 = fails.replace(
        // skip nullable production <char_null_two>
        "<char_null_one> ::= <char_null_two>",
        "<char_null_one> ::= <char_null_three>",
    );
    assert!(passes_1.parse::<Grammar>().unwrap().parse_input(input).next().is_some());

    let passes_2 = fails.replace(
        // replace <whitespace> with its terminal ' '
        "<ws> ::= <whitespace> | ' ' <ws>",
        "<ws> ::= ' ' | ' ' <ws>",
    );
    assert!(passes_2.parse::<Grammar>().unwrap().parse_input(input).next().is_some());

    let passes_3 = fails.replace(
        // again, replace <whitespace> with its terminal ' '
        "<special-char> ::= <char> | <whitespace>",
        "<special-char> ::= <char> | ' '",
    );
    assert!(passes_3.parse::<Grammar>().unwrap().parse_input(input).next().is_some());

    assert!(
        fails.parse::<Grammar>().unwrap().parse_input(input).next().is_some(),
        "all grammars except last one parsed: {input}"
    );
}
```

### PR DESCRIPTION
**Pull request by CrockAgile**
Closes #117 and #118 (hopefully!)

#117 and #118 raised examples of failures to parse input based on grammars. after some investigation, it _seems_ the root cause was shared by both issues.

## Root Cause

Consider the grammar:

```text
<start> ::= <shortfail> | <longsuccess>
<shortfail> ::= <char> 'never'
<char> ::= 'a'
<longsuccess> ::= <long2>
<long2> ::= <long3>
<long3> ::= <long4>
<long4> ::= <char>
```

When parsing input "a", there are two routes: "shortfail" and "longsuccess". As the names suggest, "shortfail" requires fewer traversals, but always fails. "longsuccess" requires more traversal steps, and should eventually succeed.

The issue is caused because both paths predict the non-terminal "char". Practical Earley parsing requires de-duplicating predictions or else recursive grammars fall into infinite loops. The existing Earley implementation in BNF does roughly the following:

(work roughly alternates between the short and long routes because new traversals are appended to the end of the work queue)

```
* <start> ::= • <shortfail>
* <start> ::= • <longsuccess>
* <shortfail> ::= • <char> 'never'
* <longsuccess> ::= • <long2>
* <char> ::= • 'a'
* <long2> ::= • <long3>
* <char> ::= 'a' •
* <long3> ::= • <long4>
* <shortfail> ::= <char> • 'never' // <--- notice this does NOT succeed
* <long4> ::= • <char> // !!! this <char> prediction is IGNORED because an identical prediction was already made
```

All the `<longN>` productions are necessary because otherwise the `<longsuccess>` route is able to predict <char> before its completion.

> I am sorry if I have not explained this super clearly. It is a tricky problem! Funny aside, I actually thought of this bug while developing v0.4.0 . But I (wrongly) assumed there was something about the Earley algorithm which resolved it. I attempted to write a test, but I did not realize it would require so many intermediate non-terminals to expose. Woops!

## Solution

The underlying problem is that because non-terminals can be shared across multiple traversal routes, "completion" must be addressed also during "prediction".

### Performance

On my machine, there was a roughly 10% performance cost to the new parsing logic. I will experiment in the future for improvements, but in the meantime I believe passing tests and correct behavior outweigh the cost.

## Extra

Basically every time I have worked on an Earley bug, I have ended up adding the same manual logging to help with debugging. I decided to commit that logging this time!

There is a new `tracing::event!` which by default is a noop, but with the "tracing" feature enabled adds logging events.

For Earley, traversal state events are logged when created and during predict/scan/complete. It helps quite a bit with debugging!

### CODE REVIEW

**Comment by CrockAgile** [comment]
I plan to keep this PR open so the author of #117 and #118 (@amascolo) can experiment with the fix 🙌

File: src/grammar.rs [hunk 1]
```diff
@@ -685,6 +685,137 @@ mod tests {
...
+        assert!(
+            fails
+                .parse::<Grammar>()
+                .unwrap()
+                .parse_input(input)
+                .next()
+                .is_some(),
+            "all grammars except last one parsed: {input}"
+        );
+    }
```
Comment (CrockAgile) [inline src/grammar.rs:817]: @amascolo I believe I have included your two reproduction cases verbatim here! If you see anything amiss, please let me know. I might end up renaming some of the variables (like `fail`) now that they are passing, but wanted to leave them untouched until you were done experimenting

Comment (amascolo) [reply src/grammar.rs:817]: Hey, just saw this PR now – I'm currently travelling – will gladly experiment next week and let you know 🙂

Comment (amascolo) [reply src/grammar.rs:817]: This is a definitely a step forward.

I have tested this branch locally –  it fixes the cases I had raised, without introducing any obvious regressions.

However, I am still seeing issues with nullable productions that fail to match: https://github.com/shnewto/bnf/issues/117#issuecomment-1377411514

On the larger grammar I'm writing, also still seeing cases where changing productions from left- to right- recursive fails to match – but it's tricky to investigate until the case above is fixed, so I might wait until that's resolved.

**Comment by amascolo** [comment]
Oh something else has been bothering me – @CrockAgile's comment makes me wonder if it's related:

> Practical Earley parsing requires de-duplicating predictions or else recursive grammars fall into infinite loops.

I tried writing a simple fuzzing test, which uses `bnf` to generate random inputs and parse them:

```rs
use bnf::Grammar;

#[test]
fn fuzz() {
    let grammar: Grammar = … // the larger grammar I'm writing
    for _ in 0..1_000 {
        // TODO it's not clear why generation sometimes fails
        if let Ok(input) = grammar.generate() {
            assert!(grammar.parse_input(&input).next().is_some());
        }
    }
}
```

However, generating inputs sometime fails:

```rs 
let input = grammar.generate().unwrap();
```

If I switch to the line above, it hits a recursion error:

```rs
RecursionLimit("Limit for recursion reached processing <rule>!")
```

What I'm wondering is:

1. Are there any reasons why a call to `Grammar::generate()` should ever fail?

2. Is there a connection between generation hitting a recursion limit and parsing failing?

3. Would fuzzing tests (that ensure generation always succeeds) help catch these issues?

This might well be unrelated – just thought it was worth raising here!

**Comment by shnewto** [comment]
@amascolo wrt the `Grammar::generate()` comment, good spot! that's def something that needs some effort, though I do feel it's unrelated to the work tracked in this PR. For context it's a feature that was added quite some time ago with the main goal being, I wanted to generate poems _and_ have a fuzz/property tests that generated bnfs for this crate to parse 😄  we're tracking it in #70 and in our own tests we're basically swallowing the recursion limit errors. 

https://github.com/shnewto/bnf/blob/f3ca02101c71636be2d0e7e1605578e63767082c/tests/grammar.rs#L29-L35

The "why" of the error it is that the current implementation just naively traverses the grammar recursively, while monitoring the stack, if we make it near the stack limit, we exit in an error instead of a panic.

File: src/grammar.rs (hunk 1 above)
Comment (amascolo) [reply src/grammar.rs:817]: @CrockAgile any chance of merging this PR and releasing it as `0.4.4` please?

It's definitely an improvement over `0.4.3` as it fixes the more noticeable edge cases that are failing.

Then you can close #118, while leaving #117 open for fixing the remaining cases – which are more involved and would probably warrant a separate PR anyway.
